Test compilation failure with required relationship

##### **Overview of the issue**

> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project jhipster: Compilation failure
[ERROR] /C:/dev/workspace/jhipster/tests/mn-mvn-jwt-angular/src/test/java/com/mycompany/myapp/web/rest/OperationResourceIT.java:[84,35] variable label might not have been initialized

##### **Reproduce the error**

Generate entities with a `required` relationshipValidateRules.

##### **Related issues**

<!-- Has a similar issue been reported before? Please search both closed & open issues -->

##### **Suggest a Fix**

<!-- If you can't fix the bug yourself, perhaps you can point to what might be
  causing the problem (line of code or commit) -->

##### **Project configuration**

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

```
entity BankAccount {
  name String required,
  balance BigDecimal required
}
entity Label {
  label String required minlength(3)
}
entity Operation {
  date Instant required,
  description String,
  amount BigDecimal required
}
relationship OneToMany {
  BankAccount{operation} to Operation{bankAccount(name)}
}
relationship ManyToOne {
  BankAccount{user(login)} to User
}
relationship ManyToMany {
  Operation{label(label) required} to Label{operation}
}
```

##### **Browsers and Operating System**

<!-- What OS are you on? is this a problem with all browsers or only IE8? -->

-   [x] Checking this box is mandatory (this is just to show you read everything)

--- a/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
+++ b/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
@@ -78,6 +78,10 @@ import io.micronaut.http.client.RxHttpClient;
 import io.micronaut.http.client.annotation.Client;
 import io.micronaut.http.client.exceptions.HttpClientResponseException;
 import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
+<%_ if (databaseType === 'sql') { _%>
+import io.micronaut.transaction.SynchronousTransactionManager;
+import io.micronaut.transaction.TransactionOperations;
+<%_ } _%>
 
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
@@ -88,6 +92,9 @@ import org.junit.jupiter.api.extension.ExtendWith;
 <%_ } _%>
 
 import javax.inject.Inject;
+<%_ if (databaseType === 'sql' && !reactive) { _%>
+import javax.persistence.EntityManager;
+<%_ } _%>
 <% if (fieldsContainBigDecimal === true) { %>
 import java.math.BigDecimal;<% } %><% if (fieldsContainBlob === true && databaseType === 'cassandra') { %>
 import java.nio.ByteBuffer;<% } %><% if (reactiveRepositories || fieldsContainDuration === true) { %>
@@ -100,6 +107,9 @@ import java.time.ZoneId;<% } %><% if (fieldsContainInstant === true) { %>
 import java.time.temporal.ChronoUnit;<% } %>
 <%_ if (searchEngine === 'elasticsearch') { _%>
 import java.util.Collections;
+<%_ } _%>
+<%_ if (databaseType === 'sql') { _%>
+import java.sql.Connection;
 <%_ } _%>
 import java.util.List;
 <%_ if (databaseType === 'cassandra' || fieldsContainUUID === true || (databaseType === 'sql' && hasOauthUser === true)) { _%>
@@ -309,6 +319,14 @@ public class <%= entityClass %>ResourceIT <% if (databaseType === 'cassandra') {
     <%_ } _%>
     @Inject
     private <%= entityClass %>Repository <%= entityInstance %>Repository;
+    <%_ if (databaseType === 'sql') { _%>
+
+    @Inject
+    private EntityManager em;
+
+    @Inject
+    SynchronousTransactionManager<Connection> transactionManager;
+    <%_ } _%>
     <%_ if (isUsingMapsId === true && ( dto !== 'mapstruct' && service === 'no')) { _%>
     @Inject
     private <%= mapsIdEntity %>Repository <%= mapsIdRepoInstance %>;
@@ -331,12 +349,16 @@ public class <%= entityClass %>ResourceIT <% if (databaseType === 'cassandra') {
 
     @BeforeEach
     public void initTest() {
-        <%= asEntity(entityInstance) %> = createEntity();
+        <%= asEntity(entityInstance) %> = createEntity(transactionManager, em);
     }
 
     @AfterEach
     public void cleanUpTest() {
+        <%_ if (databaseType === 'sql' && !reactive) { _%>
+        deleteAll(transactionManager, em);
+        <%_ } else { _%>
         <%= asEntity(entityInstance) %>Repository.deleteAll();
+        <%_ } _%>
     }
 
     /**
@@ -345,7 +367,7 @@ public class <%= entityClass %>ResourceIT <% if (databaseType === 'cassandra') {
      * This is a static method, as tests for other entities might also need it,
      * if they test an entity which requires the current entity.
      */
-    public static <%= asEntity(entityClass) %> createEntity() {
+    public static <%= asEntity(entityClass) %> createEntity(<% if (databaseType === 'sql') { %>TransactionOperations<Connection> transactionManager, EntityManager em<% } %>) {
         <%_ if (fluentMethods) { _%>
         <%= asEntity(entityClass) %> <%= asEntity(entityInstance) %> = new <%= asEntity(entityClass) %>()<% for (idx in fields) { %>
             .<%= fields[idx].fieldName %>(<%= "DEFAULT_" + fields[idx].fieldNameUnderscored.toUpperCase() %>)<% if ((fields[idx].fieldType === 'byte[]' || fields[idx].fieldType === 'ByteBuffer') && fields[idx].fieldTypeBlobContent !== 'text') { %>
@@ -373,16 +395,32 @@ public class <%= entityClass %>ResourceIT <% if (databaseType === 'cassandra') {
         // Add required entity
             <%_ if (alreadyGeneratedEntities.indexOf(otherEntityName) == -1) { _%>
     <%_ if (otherEntityName === 'user') { // TODO or other entity has no unique fields _%>
-        <%= asEntity(otherEntityNameCapitalized) %> <%= otherEntityName %> = <%= otherEntityNameCapitalized %>ResourceIT.createEntity();
+        <%= asEntity(otherEntityNameCapitalized) %> <%= otherEntityName %> = <%= otherEntityNameCapitalized %>ResourceIT.createEntity(transactionManager);
+                <%_ if (databaseType === 'sql' && !reactive) { _%>
+        transactionManager.executeWrite(status -> {
+                em.persist(<%= otherEntityName %>);
+                em.flush();
+                return <%= otherEntityName %>;
+        });
+                <%_ } _%>
                 <%_ if (databaseType === 'mongodb') { _%>
         <%= otherEntityName %>.setId("fixed-id-for-tests");
                 <%_ } _%>
     <%_ } else { _%>
         <%= asEntity(otherEntityNameCapitalized) %> <%= otherEntityName %>;
-        <%_ if (databaseType === 'sql') { _%>
-
+        <%_ if (databaseType === 'sql' && !reactive) { _%>
+        if (TestUtil.findAll(transactionManager, em, <%= otherEntityNameCapitalized %>.class).isEmpty()) {
+            <%= otherEntityName %> = <%= otherEntityNameCapitalized %>ResourceIT.createEntity(transactionManager, em);
+            transactionManager.executeWrite(status -> {
+                em.persist(<%= otherEntityName %>);
+                em.flush();
+                return <%= otherEntityName %>;
+            });
+        } else {
+            <%= otherEntityName %> = TestUtil.findAll(transactionManager, em, <%= otherEntityNameCapitalized %>.class).get(0);
+        }
         <%_ } else { _%>
-        <%= otherEntityName %> = <%= otherEntityNameCapitalized %>ResourceIT.createEntity());
+        <%= otherEntityName %> = <%= otherEntityNameCapitalized %>ResourceIT.createEntity(<% if (databaseType === 'sql') { %>transactionManager, em<% } %>));
         <%_ } _%>
         <%_ if (databaseType === 'mongodb') { _%>
         <%= otherEntityName %>.setId("fixed-id-for-tests");
@@ -398,6 +436,32 @@ public class <%= entityClass %>ResourceIT <% if (databaseType === 'cassandra') {
         <%_ } } _%>
         return <%= asEntity(entityInstance) %>;
     }
+    <%_ if (databaseType === 'sql' && !reactive) { _%>
+
+    /**
+     * Delete all <%= asEntity(entityInstance) %> entities.
+     *
+     * This is a static method, as tests for other entities might also need it,
+     * if they test an entity which requires the current entity.
+     */
+    public static void deleteAll(TransactionOperations<Connection> transactionManager, EntityManager em) {
+        TestUtil.removeAll(transactionManager, em, <%= entityClass %>.class);
+        <%_
+            const alreadyDeletedEntities = []
+            for (idx in relationships) {
+                const relationshipValidate = relationships[idx].relationshipValidate;
+                const otherEntityName = relationships[idx].otherEntityName;
+                const otherEntityNameCapitalized = relationships[idx].otherEntityNameCapitalized;
+                const mapsIdUse = relationships[idx].useJPADerivedIdentifier;
+                if ((relationshipValidate !== null && relationshipValidate === true) || mapsIdUse === true) { _%>
+        // Delete required entities
+                <%_ if (alreadyDeletedEntities.indexOf(otherEntityName) == -1) { _%>
+        <%= otherEntityNameCapitalized %>ResourceIT.deleteAll(transactionManager, em);
+                <%_ } _%>
+            <%_ } _%>
+        <%_ } _%>
+    }
+    <%_ } _%>
 
 <%_ if (!readOnly) { _%>
 

--- a/generators/server/templates/src/test/java/package/web/rest/TestUtil.java.ejs
+++ b/generators/server/templates/src/test/java/package/web/rest/TestUtil.java.ejs
@@ -3,13 +3,22 @@ package <%=packageName%>.web.rest;
 import com.fasterxml.jackson.annotation.JsonInclude;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+<%_ if (databaseType === 'sql') { _%>
+import io.micronaut.transaction.TransactionOperations;
+<%_ } _%>
 
+<%_ if (databaseType === 'sql') { _%>
 import javax.persistence.EntityManager;
 import javax.persistence.TypedQuery;
 import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaDelete;
 import javax.persistence.criteria.CriteriaQuery;
 import javax.persistence.criteria.Root;
+<%_ } _%>
 import java.io.IOException;
+<%_ if (databaseType === 'sql') { _%>
+import java.sql.Connection;
+<%_ } _%>
 import java.util.List;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -72,21 +81,43 @@ public final class TestUtil {
         assertThat(domainObject1.hashCode()).isEqualTo(domainObject2.hashCode());
     }
 
-    /**
+<%_ if (databaseType === 'sql') { _%>
+/**
      * Makes a an executes a query to the EntityManager finding all stored objects.
      * @param <T> The type of objects to be searched
+     * @param transactionManager The instance of the TransactionOperations
      * @param em The instance of the EntityManager
      * @param clss The class type to be searched
      * @return A list of all found objects
      */
-    public static <T> List<T> findAll(EntityManager em, Class<T> clss) {
-        CriteriaBuilder cb = em.getCriteriaBuilder();
-        CriteriaQuery<T> cq = cb.createQuery(clss);
-        Root<T> rootEntry = cq.from(clss);
-        CriteriaQuery<T> all = cq.select(rootEntry);
-        TypedQuery<T> allQuery = em.createQuery(all);
-        return allQuery.getResultList();
+    public static <T> List<T> findAll(TransactionOperations<Connection> transactionManager, EntityManager em, Class<T> clss) {
+        return transactionManager.executeRead(status -> {
+            CriteriaBuilder cb = em.getCriteriaBuilder();
+            CriteriaQuery<T> cq = cb.createQuery(clss);
+            Root<T> rootEntry = cq.from(clss);
+            CriteriaQuery<T> all = cq.select(rootEntry);
+            TypedQuery<T> allQuery = em.createQuery(all);
+            return allQuery.getResultList();
+        });
     }
 
+    /**
+     * Makes a an executes a query to the EntityManager removing all stored objects.
+     * @param transactionManager The instance of the TransactionOperations
+     * @param em The instance of the EntityManager
+     * @param clss The class type to be searched
+     * @param <T>
+     */
+    public static <T> void removeAll(TransactionOperations<Connection> transactionManager, EntityManager em, Class<T> clss) {
+        transactionManager.executeWrite(status -> {
+            CriteriaBuilder cb = em.getCriteriaBuilder();
+            CriteriaDelete<T> query = cb.createCriteriaDelete(clss);
+            Root<T> root = query.from(clss);
+            query.where(root.isNotNull());
+            return em.createQuery(query).executeUpdate();
+        });
+    }
+
+<%_ } _%>
     private TestUtil() {}
 }

--- a/test-integration/samples/mvn-no-client.jdl
+++ b/test-integration/samples/mvn-no-client.jdl
@@ -51,13 +51,13 @@ entity Operation {
   amount BigDecimal required
 }
 relationship OneToMany {
-  BankAccount{operation} to Operation{bankAccount(name)}
+  BankAccount{operation} to Operation{bankAccount(name) required}
 }
 relationship ManyToOne {
   BankAccount{user(login)} to User
 }
 relationship ManyToMany {
-  Operation{label(label)} to Label{operation}
+  Operation{label(label)} to Label{operation required}
 }
 
 paginate Operation with infinite-scroll

